storage/scylla_api.c

```diff
@@ -0,0 +1,606 @@
+  return rep;
+}
+void free_Transaction(Transaction** obj) {
+  free((*obj)->address);
+  free((*obj)->bundle);
+  free((*obj)->transaction);
+  free((*obj)->trunk);
+  free((*obj)->branch);
+  free((*obj)->message);
+  free(*obj);
+  return;
+}
+
+static const Select_Query select_query[NUM_OF_SELECT_METHODS] = {
+    {"SELECT * FROM permanent.bundleTable WHERE bundle = ?"},
+    {"SELECT * FROM permanent.bundleTable WHERE bundle = ? AND address = ?"}};
+
+static void print_error(CassFuture* future) {
+  const char* message;
+  size_t message_length;
+  cass_future_error_message(future, &message, &message_length);
+  ta_log_error("Error: %.*s\n", (int)message_length, message);
+}
+
+CassCluster* create_cluster(const char* hosts) {
+  CassCluster* cluster = cass_cluster_new();
+  cass_cluster_set_contact_points(cluster, hosts);
+  return cluster;
+}
+
+CassError connect_session(CassSession* session, const CassCluster* cluster) {
+  CassError rc = CASS_OK;
+  CassFuture* future = cass_session_connect(session, cluster);
+
+  cass_future_wait(future);
+  rc = cass_future_error_code(future);
+  if (rc != CASS_OK) {
+    print_error(future);
+  }
+  cass_future_free(future);
+
+  return rc;
+}
+
+CassError execute_query(CassSession* session, const char* query) {
+  CassError rc = CASS_OK;
+  CassStatement* statement = cass_statement_new(query, 0);
+  CassFuture* future = cass_session_execute(session, statement);
+
+  cass_future_wait(future);
+
+  rc = cass_future_error_code(future);
+  if (rc != CASS_OK) {
+    print_error(future);
+  }
+
+  cass_future_free(future);
+  cass_statement_free(statement);
+
+  return rc;
+}
+
+CassError prepare_query(CassSession* session, const char* query, const CassPrepared** prepared) {
+  CassError rc = CASS_OK;
+  CassFuture* future = NULL;
+
+  future = cass_session_prepare(session, query);
+  cass_future_wait(future);
+
+  rc = cass_future_error_code(future);
+  if (rc != CASS_OK) {
+    print_error(future);
+  } else {
+    *prepared = cass_future_get_prepared(future);
+  }
+
+  cass_future_free(future);
+
+  return rc;
+}
+
+CassError execute_statement(CassSession* session, CassStatement* statement) {
+  CassError rc = CASS_OK;
+  CassFuture* future = NULL;
+  future = cass_session_execute(session, statement);
+
+  cass_future_wait(future);
+
+  rc = cass_future_error_code(future);
+  if (rc != CASS_OK) {
+    print_error(future);
+  }
+
+  cass_future_free(future);
+  cass_statement_free(statement);
+
+  return rc;
+}
+static status_t mallocAndCopyBlob(cass_byte_t** target, const cass_byte_t* src, size_t size) {
+  if (src == NULL || size == 0) {
+    return SC_STORAGE_INVAILD_INPUT;
+  }
+
+  *target = malloc(sizeof(cass_byte_t) * (size + 1));
+  if (*target == NULL) {
+    ta_log_error("%s\n", "malloc failure");
+    return SC_STORAGE_OOM;
+  }
+  memcpy(*target, src, size);
+  *((*target) + size) = '\0';
+
+  return SC_OK;
+}
+
+static status_t mallocAndCopyText(char** target, const char* src, size_t size) {
```

Comment:
Why did this function be used only once?

Reply:
I've removed this function cause we don't need malloc space to each struct member now.